### ISSUE TITLE
Using graculus with edge_attr instead of pos

### ISSUE THREAD

**Issue by Matimath** [pyg-team/pytorch_geometric#873]
Hi @rusty1s , lets say that I want to replicate ChebNet paper with Mnist classification. In that paper edge_weight is defined as top 8 of correlations - namely I have a custom weight for each edge and I do not have pos attribute. My question is how can I use graculus in this situation?

I assume that I should store my custom weights in edge_attr giving me data of this form:
Data(edge_attr=[6272], edge_index=[2, 6272], x=[784, 1], y=[1])
 
Unfortunately it seems like Batch primitive does not support edge_attr. When I do:
`

        data.x = F.relu(self.conv1(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut(edge_index, edge_weight)
        cluster = graclus(data.edge_index, weight, num_nodes=784)
        data.batch=None
        return max_pool(cluster, data)`

Where data is an instance of Data, as a result I get an instance of Batch (which does not contain edge_attr that I could use for subsequent graculus). Could you tell me how to solve it? I would be very gratefull if you told me how to rewrite this:
`

    def forward(self, data):
        data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        data.edge_attr = None
        data = max_pool(cluster, data, transform=transform)

        data.x = F.elu(self.conv2(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        x, batch = max_pool_x(cluster, data.x, data.batch)

        x = global_mean_pool(x, batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        return F.log_softmax(self.fc2(x), dim=1)`

so that it uses edge_attr instead of data.pos (assuming convolutions here are just ChebConv's).

I wish you best during christmass time.
Thanks.

**Comment by rusty1s** [comment]
`Batch` should naturally support edge attributes/weights. Are they missing in the `batch` object when using the `DataLoader`?

When using Graclus in combination with edge weights, it will greedily cluster node pairs based on maximum edge weight. In your case, you can directly use `graclus` in combination with your 1d edge attributes:
```python
data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
cluster = graclus(data.edge_index, data.edge_attr, data.x.size(0))
# No need to reset edge attributes here. It will be the sum of the union of edge weights.
data = max_pool(cluster, data, transform=transform)
```

**Comment by Matimath** [comment]
Thank you very much for your response. It seems like I have done some
mistake, but your proposed solution is very helpful for me.
Thanks again for this amazing framework.
Best,
Mateusz

pt., 27 gru 2019 o 11:16 Matthias Fey <notifications@github.com> napisał(a):
